### PR TITLE
Balena s3 compressed source

### PR DESCRIPTION
**Pull request by zvin**
To be used in the image-maker

### CODE REVIEW

File: lib/source-destination/balena-s3-source.ts [hunk 1]
```diff
@@ -27,42 +28,111 @@ import { ZipSource } from './zip';
...
+	private static filesMissingFromPreloadedImages = [
+		'VERSION',
+		'VERSION_HOSTOS',
+		'device-type.json',
+	];
+
+	constructor({
+		host = 'https://s3.amazonaws.com',
+		bucket = 'resin-production-img-cloudformation',
+		prefix = 'images',
```
Comment (dfunckt) [inline lib/source-destination/balena-s3-source.ts:63]: Can you please remove these defaults? These options are already defaulted at an earlier point and I'd rather we ensured these values were explicitly pushed down here.

File: lib/source-destination/balena-s3-source.ts [hunk 2]
```diff
@@ -27,42 +28,111 @@ import { ZipSource } from './zip';
...
+	protected getUrl(path: string): string {
+		let prefix = this.prefix;
+		let release = this.release;
+		// Preloaded images have no VERSION, VERSION_HOSTOS or device-type.json file, we need to get it from the images or esr-images folder
+		if (
+			release !== undefined &&
+			BalenaS3SourceBase.filesMissingFromPreloadedImages.includes(path)
+		) {
+			release = undefined;
+			prefix = this.isESR() ? 'esr-images' : 'images';
```
Comment (dfunckt) [inline lib/source-destination/balena-s3-source.ts:101]: Can you please move these two constants to proper module-level constants?

File: examples/balena-s3-compressed-configure.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+		yargs.option('host', {
+			type: 'string',
+			default: 'https://s3.amazonaws.com',
+			description: 'S3 host including scheme and port',
+		});
+		yargs.option('bucket', {
+			type: 'string',
+			default: 'resin-production-img-cloudformation',
+		});
+		yargs.option('prefix', { type: 'string', default: 'images' });
```
Comment (dfunckt) [inline examples/balena-s3-compressed-configure.ts:108]: Is *not* providing defaults for these options less desirable? What is the reason? I'd personally rather the SDK assumed the least number of things possible.

File: lib/source-destination/balena-s3-source.ts (hunk 1 above)
Comment (zvin) [reply lib/source-destination/balena-s3-source.ts]: Removed

File: lib/source-destination/balena-s3-source.ts (hunk 2 above)
Comment (zvin) [reply lib/source-destination/balena-s3-source.ts]: Added, but these values aren't used anywhere else.

File: examples/balena-s3-compressed-configure.ts (hunk 3 above)
Comment (zvin) [reply examples/balena-s3-compressed-configure.ts:108]: This is in the `examples` folder, it is not really part of the sdk.
It is intended as documentation or as a quick way to test the code.

File: lib/zip.ts [hunk 4]
```diff
@@ -41,7 +41,14 @@ export const getFileStreamFromZipStream = async (
...
 						// Stop reading the zip archive once the file we want has been extracted.
-						zipStream.unpipe(unzipper);
+						if (zipStream.unpipe !== undefined) {
+							zipStream.unpipe(unzipper);
+						}
+						// @ts-ignore
+						if (zipStream.destroy !== undefined) {
+							// @ts-ignore
+							zipStream.destroy();
+						}
```
Comment (Page-) [inline lib/zip.ts:51]: You can do this in one:
```suggestion
						zipStream.destroy?.();
```

File: lib/zip.ts [hunk 5]
```diff
@@ -41,7 +41,14 @@ export const getFileStreamFromZipStream = async (
 					}
 					entry.on('end', () => {
 						// Stop reading the zip archive once the file we want has been extracted.
-						zipStream.unpipe(unzipper);
+						if (zipStream.unpipe !== undefined) {
+							zipStream.unpipe(unzipper);
+						}
```
Comment (Page-) [inline lib/zip.ts:46]: You can do this in one:
```suggestion
						zipStream.unpipe?.(unzipper);
```

File: lib/stream-limiter.ts [hunk 6]
```diff
@@ -37,7 +37,9 @@ export class StreamLimiter extends Transform {
 		}
 		this.maxBytes -= length;
 		if (this.maxBytes === 0) {
-			this.stream.unpipe(this);
+			if (this.stream.unpipe !== undefined) {
+				this.stream.unpipe(this);
+			}
```
Comment (Page-) [inline lib/stream-limiter.ts:42]:
```suggestion
			this.stream.unpipe?.(this);
```

File: examples/data/ts4900.config.json [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+	"registryEndpoint": "registry2.resinstaging.io",
+	"userId": 392,
+	"username": "alexis",
+	"version": "2.9.6+rev1.prod",
+	"vpnEndpoint": "vpn.resinstaging.io",
+	"vpnPort": 443,
+	"wifiKey": "pass",
+	"wifiSsid": "ssid",
+	"processorCore": "quad"
+}
```
Comment (Page-) [inline examples/data/ts4900.config.json:22]: These don't exist in modern configs:
```suggestion
{
	"apiEndpoint": "https://api.resinstaging.io",
	"apiKey": "nXasvWVTDc5B3YKITzdkcDbHdcef9uzH",
	"appUpdatePollInterval": 2520000,
	"applicationId": 20605,
	"deltaEndpoint": "https://delta.resinstaging.io",
	"deviceType": "raspberrypi3",
	"listenPort": 48484,
	"mixpanelToken": "cb974f32bab01ecc1171937026774b18",
	"registryEndpoint": "registry2.resinstaging.io",
	"version": "2.9.6+rev1.prod",
	"vpnEndpoint": "vpn.resinstaging.io",
	"vpnPort": 443,
	"wifiKey": "pass",
	"wifiSsid": "ssid",
	"processorCore": "quad"
}
```

File: examples/data/rpi3.config.json [hunk 8]
```diff
@@ -1 +1,21 @@
...
+	"pubnubSubscribeKey": "sub-c-bbc12eba-ce4a-11e3-9782-02ee2ddab7fe",
+	"registryEndpoint": "registry2.resinstaging.io",
+	"userId": 392,
+	"username": "alexis",
+	"version": "2.9.6+rev1.prod",
+	"vpnEndpoint": "vpn.resinstaging.io",
+	"vpnPort": 443,
+	"wifiKey": "pass",
+	"wifiSsid": "ssid"
+}
```
Comment (Page-) [inline examples/data/rpi3.config.json:21]: These don't exist in modern configs:
```suggestion
{
	"apiEndpoint": "https://api.resinstaging.io",
	"apiKey": "nXasvWVTDc5B3YKITzdkcDbHdcef9uzH",
	"appUpdatePollInterval": 2520000,
	"deltaEndpoint": "https://delta.resinstaging.io",
	"deviceType": "raspberrypi3",
	"listenPort": 48484,
	"mixpanelToken": "cb974f32bab01ecc1171937026774b18",
	"registryEndpoint": "registry2.resinstaging.io",
	"version": "2.9.6+rev1.prod",
	"vpnEndpoint": "vpn.resinstaging.io",
	"vpnPort": 443,
	"wifiKey": "pass",
	"wifiSsid": "ssid"
}
```

Comment (zvin) [reply examples/data/rpi3.config.json]: updated

File: examples/data/ts4900.config.json (hunk 7 above)
Comment (zvin) [reply examples/data/ts4900.config.json]: updated

**Review by Page- — APPROVED** [review]

**Pull request merged**